File: python-backend/src/application/dtos/document_dtos.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import List, Optional

# -----------------------------------------------------------------------------
# Data Transfer Object (DTO) for Document Metadata/API Response
#
# This DTO is used to define the structure of document records that are 
# returned to the client via API responses. It provides both file metadata and 
# processing results, serving as the authoritative schema for document-related 
# exchanges in the system.
# -----------------------------------------------------------------------------
class DocumentResponse(BaseModel):
    """
    Represents the API response structure for a document, including uploaded file 
    metadata, processing status, any extracted analysis, and validation results.

    Attributes:
        id (int): Unique identifier for the document.
        file_name (str): Name of the uploaded file (serialization alias: "fileName").
        content_type (str): MIME type of the file (serialization alias: "contentType").
        file_size (int): File size in bytes (serialization alias: "fileSize").
        is_processed (bool): Whether this document has been processed/analyzed 
            (serialization alias: "isProcessed").
        analysis_result (Optional[str]): The textual or JSON result of document 
            analysis (nullable, serialization alias: "analysisResult").
        creation_date (datetime): Date/time when the document was uploaded 
            (serialization alias: "creationDate").
        validation_errors (Optional[List[str]]): Any errors encountered during
            validation or processing, if applicable (nullable, serialization alias: "validationErrors").
    """
    id: int  # Unique identifier for the document (primary key)
    file_name: str = Field(serialization_alias="fileName")  # Name of the uploaded file
    content_type: str = Field(serialization_alias="contentType")  # MIME type of file
    file_size: int = Field(serialization_alias="fileSize")  # Size of file in bytes
    is_processed: bool = Field(serialization_alias="isProcessed")  # True if document has been processed
    analysis_result: Optional[str] = Field(default=None, serialization_alias="analysisResult")  # Analysis or parsing output, if available
    creation_date: datetime = Field(serialization_alias="creationDate")  # Upload timestamp
    validation_errors: Optional[List[str]] = Field(default=None, serialization_alias="validationErrors")  # Validation or business rule errors, if any

    class Config:
        # Allow population of model fields from ORM object attributes
        from_attributes = True
        # Allow population by both field names and serialization aliases for flexibility
        populate_by_name = True

File: dotnet-backend/src/Application/DTOs/DocumentDtos.cs
namespace Application.DTOs;

/// <summary>
/// Represents the request data required to upload a document.
/// </summary>
/// <param name="FileName">
/// The name of the document file being uploaded (including extension).
/// </param>
/// <param name="Content">
/// The content stream of the file. This should contain the raw file bytes.
/// </param>
/// <param name="ContentType">
/// The MIME type of the uploaded file (e.g., "application/pdf").
/// </param>
/// <param name="ProcessType">
/// (Optional) The process type or document type hint (e.g., "Invoice" or "Information").
/// </param>
public record DocumentUploadRequest(
    string FileName,
    Stream Content,
    string ContentType,
    string? ProcessType = null
);

/// <summary>
/// Represents the response for a document, including its metadata and analysis state.
/// </summary>
/// <param name="Id">
/// The unique identifier assigned to the document in the system.
/// </param>
/// <param name="FileName">
/// The document's filename (including extension).
/// </param>
/// <param name="ContentType">
/// The MIME type of the document.
/// </param>
/// <param name="FileSize">
/// The size of the file in bytes.
/// </param>
/// <param name="IsProcessed">
/// Whether the document has undergone processing/analysis.
/// </param>
/// <param name="AnalysisResult">
/// Serialized JSON string of the analysis result if processed; otherwise <c>null</c>.
/// </param>
/// <param name="CreationDate">
/// The UTC date and time when the document was uploaded/created.
/// </param>
/// <param name="ValidationErrors">
/// (Optional) List of validation errors, if the document failed validation during upload or processing.
/// Will be <c>null</c> or empty if there are no errors.
/// </param>
public record DocumentResponse(
    int Id,
    string FileName,
    string ContentType,
    long FileSize,
    bool IsProcessed,
    string? AnalysisResult,
    DateTime CreationDate,
    List<string>? ValidationErrors = null
);
